### DISCUSSION TITLE
Docs: Build from `stable/a.b.c` branches directly instead of final version tags

### DISCUSSION THREAD

**Discussion by laymonage** [wagtail/wagtail#12464] [Development]
Hi all!

The versioning (thus also the builds) for our docs are currently configured this way:
- `latest`: built from the `main` branch, i.e. a new commit triggers a new build
- `stable`: built from the latest `v`-prefixed tag ("latest" according to SemVer rules, excluding prereleases)
- `vA.B.C`: built from all `v`-prefixed tags that adhere to SemVer.

I believe this is all the default behaviour from Read the Docs (RTD) as per https://docs.readthedocs.io/en/stable/versions.html and not something we customised. The only "custom" thing/manual work we do is that we mark old patch versions as **"Hidden"** in the RTD dashboard. These versions are still accessible by URL, but not shown in the RTD flyout and not indexed by search engines. For example: https://docs.wagtail.org/en/v6.2.1/.

There might also be some older versions that we mark as **"Inactive"**, which completely deletes the docs version.

While this setup has served us quite well, there are a few pain points:

- Apart from the `latest` version, docs are only built when we push a new tag, which generally happens when creating a new release. This means we cannot update the docs without also making a Wagtail package release.
  - This is especially bad if we need to fix a security issue in our older docs. We'd like to keep the old docs around, but we might not be able to if there was a serious issue and we couldn't update it. While the docs website is static, it does have some dynamic components injected by RTD which might affect our assessment of security vulnerabilities. See https://github.com/wagtail/wagtail/issues/12444#issuecomment-2435384977 for more details.
  - This is also bad if we need to clarify the upgrade process for an older release that is no longer supported. We have seen issues with e.g. bad migration files in #10724 and #10895. While we can no longer issue a new release for people upgrading their projects, it would be nice if we could update the docs. One might argue that people should be using the `latest` version of the docs to read the release notes when upgrading, but I can see cases where this may be useful, e.g. it's in a page that has been removed.
  - This is also bad when we need to update the theme (https://github.com/wagtail/sphinx-wagtail-theme) due to changes in RTD's behaviour. For example, they recently rolled out a new flyout, which seems to have changed how the ads are rendered, and is incompatible with our theme:  <details><summary>Screenshot</summary><img src="https://github.com/user-attachments/assets/fe4438e0-5dee-48d0-b1ca-eec756cb7959" alt="ads overlapping with the table of contents" /></details>
- When creating a new patch release, we need to hide the previous patch version in RTD. Ref: https://github.com/wagtail/wagtail/wiki/Creating-a-new-Wagtail-release#docs.
- If people use version-specific links, it points to the specific _patch_ version. This means the link will be outdated when a new patch release is out, which may include documentation fixes for the linked page. They will need to update the link to the latest patch version.
  - Reason why people may do this: some docs may be removed in newer versions, e.g. the ModelAdmin to Snippets migration guide https://docs.wagtail.org/en/v5.2.3/reference/contrib/modeladmin/migrating_to_snippets.html
- Our docs versions on RTD become "bloated". We currently have **127** versions on RTD, and this number will grow with _every_ patch release. See https://app.readthedocs.org/projects/wagtail/.

Therefore, I'd like to make a proposal to build our docs directly from our `stable/A.B.x` branches instead of the release tags. This entails a few changes to future documentation updates and Wagtail releases.

1. The docs versions on RTD will become `vA.B.x`, e.g. `v6.3.x` rather than specific patch versions like `v6.3.1`.
   - This is reflected in the version selector inside the RTD flyout at the bottom right, as well as the URLs, e.g. https://docs.wagtail.org/en/v6.2.x/
   - This applies to future versions. TBD: do we want to do this for past versions too?
2. Every push to the `stable/A.B.x` branch will trigger a docs rebuild for the `vA.B.x` docs version, without having to wait for a new `vA.B.x` release.

## Downsides

- When a patch release is still under development, i.e. it's in an "Unreleased" state, the corresponding `vA.B.x` docs will already have those changes. This may cause confusion to users, e.g. they might think the newly-documented behaviour is applicable to their installed version.
  - I think this is rare though, and we could add the appropriate `versionadded`/`versionchanged` admonitions with the patch versions where necessary.

Will add more if I can think of other downsides.

## ⚠️ To be decided

There are a few things that need to be decided as part of this change.

1. Are we happy with this, i.e. building directly from the `stable/A.B.x` branch rather than the final tagged version?
   If we are, what URL slug do we want to use? Currently we use `vA.B.C`. Choose between:
     - `stable-6.3.x` (the default)
     - `v6.3.x`
     - `v6.2`
     - with or without the `v`?
     Note that the **`x`** is a literal "x" and not to be replaced with the current latest version. For anything other than the default, we need to contact RTD every time we make a new RC release.
2. Do we want to also build the docs from the `stable/A.B.x` branches of old versions, or just new ones from this point forward?
   - Reason why we might want to do this is to fix issues mentioned above, e.g. security vulnerabilities and/or issues with the RTD flyout.
4. Do we want to keep every single old patch version on RTD available? For example, we have https://docs.wagtail.org/en/v6.2.1/ which has been superseded by https://docs.wagtail.org/en/v6.2.2/ (and possibly v6.2.3 later).
   - These are currently marked as **Hidden**, i.e. available only via direct URL.
   - We can also mark them as **Inactive**, which will destroy them until we reactivate it again. The reason why we might want to do this is to reduce clutter and avoid confusion when docs mistakes have been fixed in a newer patch version.
     - If we do this, ideally we should set up redirects so people who hold on to these links get redirected to the latest patch version for the `A.B.x` release.
     - If we do this **and** (1), then the redirects should go to the latest `A.B.x` branch version (instead of the latest tagged patch version).

There may be more I'm missing, but those are the only ones I can think of right now.

## Worth noting

@saevarom also suggested the possibility of hosting the old versions as a "degraded" experience without JS, so we don't have to deal with new issues. This sounds like a good idea but I'm not sure how possible it is to keep it as part of the RTD setup (and not like a completely separate site).

Also I _think_ this is how Django's docs are built too, i.e. they target the branches directly rather than the final release. For example, looking at the history for its `stable/3.1.x` branch https://github.com/django/django/commits/stable/3.1.x, there is https://github.com/django/django/commit/0b8a0296bfd30748f08021834e95cdae241686e8 which was committed _after_ the last release was made (3.1.14). Technically it would be released as part of 3.1.15, but that release never happened – yet you can see the docs changes in https://docs.djangoproject.com/en/3.1/ref/urls/ and https://django.readthedocs.io/en/3.1.x/ref/urls.html.

Django's docs are built differently though, from what I can tell they seem to be building the docs into `.json` files that are then loaded by the `djangoproject.com` project, hence it looks a bit different. For example, the versions in the URLs don't have the `.x` suffix. See https://github.com/django/djangoproject.com/blob/main/docs/views.py for how they do it. However, they also have https://django.readthedocs.io/en/stable/ which is the more traditional RTD build, and also what it looks like if you build their docs locally. Their RTD setup can be seen at https://app.readthedocs.org/projects/django/.

**Comment by gasman** [comment]
[I did some work in this direction](https://github.com/gasman/btfd) a long time back. Since RTD doesn't (or didn't?) support any scheme other than the tag-based one, my thinking was that we'd ditch RTD entirely, and run the build locally, using my script to manage a set of checkouts of each branch, and upload to a static file host such as AWS.

One stumbling block I encountered (can't remember if it was the only one...) is that you can't necessarily tell which is the latest stable release just from looking which branches exist - for example, right now in the 6.3 release candidate period `stable/6.3.x` exists, but 6.2 is the current stable version. (Although now I think about it - you _can_ infer that from the fact that a `v6.3` tag doesn't exist.)

**Reply by laymonage** [reply]
I forgot to mention that we can still keep the `stable` version pointing to the latest tagged release. In fact, this will always be the behaviour unless we specifically have a branch or tag named `stable`: https://docs.readthedocs.io/en/stable/versions.html. This change will only affect how we keep the docs for older versions.

I see that your repo's latest commit was in August 2019 and the first commit was in October 2018. RTD added automation rules in https://github.com/readthedocs/readthedocs.org/pull/5996 which was merged in November 2019 and that should allow rules with custom regex.

I've now added the following automation following the docs https://docs.readthedocs.io/en/stable/guides/automation-rules.html#adding-a-new-automation-rule to try this out:

<img width="791" alt="image" src="https://github.com/user-attachments/assets/8f95502f-a5cc-4299-a444-380380e30127">

I think it needs to be triggered somehow, though. The docs say creating a new branch will trigger it, but perhaps pushing a commit to an existing branch would suffice? I guess we'll see when we next push a commit to `stable/6.3.x`. (Or we can try pushing to an old branch or making a test one that doesn't exist e.g. `stable/5.3.x`)

See also Django's RTD: https://django.readthedocs.io/en/stable/ and https://app.readthedocs.org/projects/django

**Reply by laymonage** [reply]
Hmm, we pushed 517fa8a8324e18703aae1d3977eb8705f7fdd0e7 but it didn't trigger a new version on RTD. Maybe it only gets triggered when a new branch is created... :(

(Also, just to clarify – the above screenshot has a mistake. The `.` should be escaped as `\.`, which I already fixed before we pushed that commit.)

Maybe Django had some manual intervention for their setup. For example, it's suspicious that the new branches on https://app.readthedocs.org/projects/django/ has the `stable/` prefix, but the older ones don't – yet the slugs on the hosted docs do not have a `stable-` prefix. There are also cases where the "branch" link points to a non-existent branch on GitHub because it's missing the `stable/` prefix, e.g. 
![image](https://github.com/user-attachments/assets/813cf354-ad8a-45cd-9555-e00114a26bd0)


According to https://github.com/readthedocs/readthedocs.org/issues/5318, https://github.com/readthedocs/readthedocs.org/issues/9684, and https://github.com/readthedocs/readthedocs.org/pull/6273#pullrequestreview-301277942 it might be possible to email support to change the slug, so maybe that's what they've done.

**Reply by laymonage** [reply]
I'm starting to believe that even if this worked, the capturing group in the regex will not do anything – i.e. the build will be available under the `stable-6.3.x` slug instead of just `6.3.x`, and we'll have to contact support if we want to change the slug.

So at the end of the day, this will be the same as activating the `stable/6.3.x` branch directly from https://app.readthedocs.org/dashboard/wagtail/version/stable-6.3.x/edit/...

**Reply by laymonage** [reply]
Confirmed: see item 6 in https://docs.djangoproject.com/en/dev/internals/howto-release-django/#new-stable-branch-tasks

See also: https://github.com/readthedocs/readthedocs.org/issues/2285 and its related issues.

**Reply by thibaudcolas** [reply]
I’m hesitant for us to proceed with something that would be entirely dependent on Read the Docs support, no matter how small, so I’ve raised the issue here: [Add alias support to version automation](https://github.com/readthedocs/readthedocs.org/issues/5318).

**Reply by laymonage** [reply]
Thanks @thibaudcolas, worth noting Eric's standpoint on the feature in this PR: https://github.com/readthedocs/readthedocs.org/pull/6273#pullrequestreview-301277942

> This feels really complex, which is the reason I have wanted to not allow this in the past. I don't think we should build an incredibly complex system for handling renaming of the slug. We should either let users shoot themselves in the foot, or not allow it. I tend to err on the side of not allowing it, and having them ask us when they want it.
> This is the same conclusion we came to last time, and I still haven't heard a good reason for why we should allow users to rename things instead of just emailing support.

That was from five years ago though, so he might've changed his mind since.

**Reply by thibaudcolas** [reply]
Would there be a case for us to rename our branches as an alternative? For example `version/6.3.x` probably would still be clear for us as a branch name, and `version-6.3.x` would be longer than ideal but otherwise decent in the URL? 🤔

**Reply by laymonage** [reply]
It's an option, but if we're going that way then we might as well use `v6.3.x` as the branch name. Seems a bit controversial of a change though, but I suppose we can keep the old names just in case people rely on it.

**Comment by allcaps** [comment]
I think it's a good idea to separate the documentation build/publish process from the package build/release process. This will provide more flexibility for both. Being able to rebuild all documentation seems a big win.

Would it be a good idea to remove the documentation for unsupported versions? Perhaps this is a bit controversial? A one- or two-year grace period might be reasonable, but keeping outdated documentation around indefinitely can be more confusing than helpful. Besides, people who really _need_ old docs can read or build them from source. I believe anything before `v2.14.2` can safely be removed.

I'm not entirely convinced about the `.x` URL scheme. Why not use the highest minor version in the URL? The earlier versions might permanently redirect. Wouldn't that be clearer? We don't need a redirect for every page, just a general redirect rule. For example:

- 6.1 -> 6.1.3
- 6.1.1 -> 6.1.3
- 6.1.2 -> 6.1.3

We also might add a line to each page: `This page applies to Wagtail 6.1, 6.1.1, 6.1.2 and 6.1.3.`

I guess we also should make release notes and upgrade considerations cumulative? The `6.1.3` version needs to contain the upgrade considerations for `6.1`.

**Reply by laymonage** [reply]
> I'm not entirely convinced about the `.x` URL scheme. Why not use the highest minor version in the URL? The earlier versions might permanently redirect. Wouldn't that be clearer? We don't need a redirect for every page, just a general redirect rule.

I'd rather avoid redirects whenever possible. A few downsides off the top of my head:

- There's one intermediary response before the actual docs. Not very good for sustainability.
- When a new patch version is released, a link that previously works directly (returns a rendered HTML response) will now return a redirect instead.
- We'll have to update the redirects to the latest patch versions whenever we make a new release.

Note that we don't have to use the `.x` – I _think_ it should be possible to host them under `/vA.B/` only e.g. `/v6.1/`. I'm just suggesting the `.x` because our current versioning scheme is `/vA.B.C/`, so changing `.C` to a fixed literal `.x` isn't that big of a change. It also signifies that it applies to all `.x` patch versions – dropping the `.x` might give the impression that it's the `.0` version of the docs.

> We also might add a line to each page: `This page applies to Wagtail 6.1, 6.1.1, 6.1.2 and 6.1.3.`

Not sure why we'd need to list those specific versions, sounds like a bit too much work to implement 🤔

Although now that you mentioned it, I seem to recall us showing a banner that reminds user when they're viewing an old version of the docs. I can't see it anymore, maybe that broke at some point or we removed it. I can see how that banner would be useful.

> I guess we also should make release notes and upgrade considerations cumulative? The `6.1.3` version needs to contain the upgrade considerations for `6.1`.

I don't think we need to copy them over – we'd be duplicating docs for every patch version if we did so. That would also be prone to mistakes e.g. forgetting to update one of the release notes.

I do think it would be useful to have a note or some kind that links to the `A.B` release notes for people who are upgrading from an older `A.B` version, though.

**Comment by thibaudcolas** [comment]
Read the Docs people are working on [Add alias support to version automation #5318](https://github.com/readthedocs/readthedocs.org/issues/5318), and they now have some support for what we’d need to make branch versions work. I’ve requested they enable those features for `wagtail` and `sphinx-wagtail-theme` so we can try them out.

**Reply by laymonage** [reply]
Oh, I didn't notice that. Thanks!

**Reply by thibaudcolas** [reply]
And it’s now enabled :) @laymonage I assume we’re ok to try this out on the branch builds you had set up for `stable-6.4.x` and `stable-6.3.x`?

**Reply by laymonage** [reply]
> And it’s now enabled :) @laymonage I assume we’re ok to try this out on the branch builds you had set up for `stable-6.4.x` and `stable-6.3.x`?

Yes please!

**Reply by thibaudcolas** [reply]
I have tried and it works great, with the caveat I didn’t want to risk testing the ordering of the versions in the picker. I assume we would only adopt this if we could be sure our newly set up branch builds would show after "latest stable". Full notes in https://github.com/readthedocs/readthedocs.org/issues/5318.

**Comment by thibaudcolas** [comment]
Briefly discussed this w/ @laymonage in private, there are a few more things we’d need to figure out to adopt branches instead of tags, beyond the slugs:

- What do we do with "stable" – would it keep pointing to a tag? Would we remove it? Would it point to the latest release’s `stable/` branch?
- How do we approach this with pre-existing releases? Do we convert all the tags to branch builds? Presuming we can’t, we’d need to figure out how to combine some of our past tags builds, and some branch builds.

For both challenges I assume we can trial what’s possible on the [sphinx-wagtail-theme](https://github.com/wagtail/sphinx-wagtail-theme) project.

Edit: also need to lock our dependencies while we’re here
